src/telemetry/cellular_modem.rs

```diff
@@ -0,0 +1,103 @@
+/*
+ * This file is part of Edgehog.
+ *
+ * Copyright 2022 SECO Mind Srl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::DeviceManagerError;
+use astarte_sdk::types::AstarteType;
+use std::collections::HashMap;
+use zbus::dbus_proxy;
+use zbus::zvariant::{DeserializeDict, SerializeDict, Type};
+
+#[derive(Debug, Clone, DeserializeDict, SerializeDict, Type)]
+#[zvariant(signature = "dict")]
+pub struct ModemProperties {
+    apn: String,
+    imei: String,
+    imsi: String,
+}
+
+#[dbus_proxy(
+    interface = "io.edgehog.CellularModems1",
+    default_service = "io.edgehog.CellularModems",
+    default_path = "/io/edgehog/CellularModems"
+)]
+trait CellularModems {
+    fn list(&self) -> zbus::Result<Vec<String>>;
+    fn get(&self, id: String) -> zbus::Result<ModemProperties>;
+}
+
+pub async fn get_cellular_properties() -> Result<HashMap<String, AstarteType>, DeviceManagerError> {
+    let mut ret: HashMap<String, AstarteType> = HashMap::new();
+    if let Ok(connection) = zbus::Connection::session().await {
+        if let Ok(proxy) = CellularModemsProxy::new(&connection).await {
+            if let Ok(available_modems) = proxy.list().await {
+                for modem_id in available_modems {
+                    let modem: ModemProperties = proxy.get(modem_id.clone()).await?;
+                    ret.extend(get_modem_properties(modem_id, modem));
+                }
+            }
+        }
+    }
```

Comment:
As in the other PR, let's try to keep (at least in new code) an early return approach vs the heavily nested approach here

Reply:
As in the other PR, you could have a `get_available_modems() -> Result<Something>` function that chains the calls starting from `zbus::Connection::session` with the `?` operator and then the outer function just handles the `Ok/Err` case in a single place and either iterates on `available_modems` or returns an empty HashMap